ARTEMIS-196 Implement Consumer Priority

michaelandrepearce opened the pull request:
@franz1981 an alternative so we don't have to have a copy of CopyOnWriteArrayList, it does mean on add or remove consumer we have to invoke toArray which causes a copy, but this is not on hot path, so i think we should be good, and avoids us having to clone a jvm class.

franz1981 commented:
@michaelandrepearce Nice! Will take a look today or max tomorrow :+1:

michaelandrepearce commented:
@franz1981 just ignore class comments, theyre the originals still, ill need to change, but wanted to get to you quickly so you have chance to look over. If you think this is better ill make final tidyup bits, such as class comments and replace the real PR's branch.

michaelandrepearce commented:
@franz1981 did you get a chance to look, do you think this is better than original solution? 

Am keen to get this feature into the next release cut.

franz1981 commented:
@mochaelandrepearce sadly haven't had much time today to look into it :( 
Tomorrow I have already scheduled some time in the morning to take a look into this

michaelandrepearce commented:
@franz1981 great thanks a million :)

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/openwire/amq/QueueConsumerPriorityTest.java:
```diff
@@ -0,0 +1,65 @@
+         producer.send(msg);
+         assertNotNull("null on iteration: " + i, highConsumer.receive(1000));
+      }
+      assertNull(lowConsumer.receive(2000));
```
Would a receiveNoWait (either in or outside the loop) like the other tests be nicer than burning 2 seconds? Slow tests is a key reason eventually noone wants to runs the tests :)

michaelandrepearce replied:
this is the original test from ActiveMQ5 i was trying to keep this test as much un-touched as possible to ensure behavior is the same.

michaelandrepearce replied:
changed.

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpReceiverPriorityTest.java:
```diff
@@ -0,0 +1,90 @@
+
+      for (int i = 0; i < 5; i++) {
+         AmqpMessage message1 = receiver1.receive(250, TimeUnit.MILLISECONDS);
+         AmqpMessage message2 = receiver2.receive(250, TimeUnit.MILLISECONDS);
```
Burning 250ms twice per loop seems excessive. There is a receiveNoWait that could be used for initial verification nothing arrived, and/or a small final timed wait could be done outside the loop afterwards. Alternatively, pullImmediate() would avoid unnecessary waiting.

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpReceiverPriorityTest.java:
```diff
@@ -0,0 +1,90 @@
+      //Check messages now goto next priority receiver
+      for (int i = 0; i < 5; i++) {
+         AmqpMessage message2 = receiver2.receive(250, TimeUnit.MILLISECONDS);
+         AmqpMessage message3 = receiver3.receive(250, TimeUnit.MILLISECONDS);
```
As above.

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpReceiverPriorityTest.java:
```diff
@@ -0,0 +1,90 @@
+      AmqpSession session = connection.createSession();
+
+      Map<Symbol, Object> properties1 = new HashMap<>();
+      properties1.put(Symbol.getSymbol("priority"), 50);
```
I'd suggest creating consumers with priorities out of order (e.g highest in middle), so they arent simply registered in sequence, as otherwise a simple failure to round-robin delivery attempts (given every receiver has enough credit to receive all messages) might also lead to the expected result even without any priority handling consideration.

michaelandrepearce replied:
This is actually tested on the queueconsumerimpl test. But agree we can do same here

michaelandrepearce commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/SessionCreateConsumerMessage.java:
```diff
@@ -52,6 +57,7 @@ public String toString() {
       StringBuffer buff = new StringBuffer(getParentString());
       buff.append(", queueName=" + queueName);
       buff.append(", filterString=" + filterString);
+      buff.append(", priority=" + priority);
```
Makes sense

michaelandrepearce replied:
done

michaelandrepearce commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/SessionCreateConsumerMessage.java:
```diff
@@ -104,6 +119,11 @@ public void decodeRest(final ActiveMQBuffer buffer) {
       filterString = buffer.readNullableSimpleString();
       browseOnly = buffer.readBoolean();
       requiresResponse = buffer.readBoolean();
+      if (buffer.readableBytes() > 0) {
```
This is typical pattern used for adding safely a new field that can be either nullable or defaultable. Used many times over.

michaelandrepearce commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/client/ConsumerPriorityTest.java:
```diff
@@ -0,0 +1,298 @@
+ */
+public class ConsumerPriorityTest extends JMSTestBase {
+
+   private SimpleString queueName = SimpleString.toSimpleString("jms.consumer.priority.queue");
```
Nice i didnt know about that in the parent class. will change to use this..

michaelandrepearce replied:
now using getName ... again nice nice....

michaelandrepearce commented on artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPSessionCallback.java:
```diff
@@ -233,6 +239,11 @@ public Object createSender(ProtonServerSenderContext protonSender,
    }
 
+   private int getPriority(Map<Symbol, Object> properties) {
+      Integer value = properties == null ? null : (Integer) properties.get(PRIORITY);
```
Have changed if you can recheck to make sure i understood.

michaelandrepearce commented:
@gemmellr thanks for review, if you could recheck the AMQPSessionCallback for me, to make sure i understood you. 

As for the Openwire test case, this was a simple port over of the existing activemq5 test case as untouched as possible, i agree we could reduce the time but id rather (at least for this release) keep it the same, so we can be sure feature works for openwire same as activemq5.

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java:
```diff
@@ -0,0 +1,124 @@
+
+   @Override
+   public QueueConsumers<T> reset() {
+      if (changedIterator != null) {
```
I will just save `changedIterator` once and will perform a logic with a local value instead of volatile load twice

michaelandrepearce replied:
Nice

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java:
```diff
@@ -0,0 +1,124 @@
+   public QueueConsumers<T> reset() {
+      if (changedIterator != null) {
+         currentIterator = changedIterator;
+         changedIteratorFieldUpdater.compareAndSet(this, changedIterator, null);
```
What happen if the cas will fail? we ends with a reset() that is not nulling `changedIterator`.
Given that we are just clearing the `changedIterator` to `null` a `lazySet(this, null) is enough.

michaelandrepearce replied:
If it fails it means whilst we were reseting and changing reference to the new iterator. a consumer was added or removed and another new iterator was created (as this can occur whilsy we are iterating or resetting.) And means simply whilst this reset updated its iterator, on next reset it needs to switch over its again. We wouldnt want to lazy set to null

michaelandrepearce replied:
currentIterator is ONLY updated by reset, which should ONLY be called by the same threads operating on the ResettableIterator interface.

When add or remove of consumer occurs a new iterator is parked into the volatile changedIterator (using an atomic field updater), so the next reset can pick it up.

michaelandrepearce replied:
Thats fair comment. Ill add this.

michaelandrepearce replied:
ill mark comment resolved, if you think it needs further docs or could be further abstracted to be more clearer let me know.

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java:
```diff
@@ -0,0 +1,124 @@
+         currentIterator = changedIterator;
+         changedIteratorFieldUpdater.compareAndSet(this, changedIterator, null);
+      } else {
+         currentIterator.reset();
```
I suppose that `reset` is safe to be called just by one thread at time, if not, it would be complex because `currentIterator` could be changed indipendently

michaelandrepearce replied:
Like any iterator and iterator should only be interacted by one thread at a time.

franz1981 replied:
So it is a good reason to save using a `compareAndSet` that's not only expensive but give the false illusion that the code is thread-safe

michaelandrepearce replied:
The compare and set is because the new iterator could be created whilst we are mid resetting. The iterator is single thread usage only. Updating the collection should be thread safe and able to occur in parralell.

This is much like iterating copyonwriteareatlist. Iterator works on a snapshot view and this is single thread. Updating the collection itself can be concurrent and occur whilst someone else is iterating

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java:
```diff
@@ -0,0 +1,124 @@
+   public boolean add(T t) {
+      boolean result = consumers.add(t);
+      if (result) {
+         changedIteratorFieldUpdater.set(this, consumers.resettableIterator());
```
`lazeSet` is enough for single-writer/single-threaded semantic

michaelandrepearce replied:
see other comment

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java:
```diff
@@ -0,0 +1,124 @@
+   public boolean remove(T t) {
+      boolean result = consumers.remove(t);
+      if (result) {
+         changedIteratorFieldUpdater.set(this, consumers.resettableIterator());
```
`lazeSet` is enough for single-writer/single-threaded semantic

michaelandrepearce replied:
Agreed. But this is not hot path, and if anything would allow in future for consumers to be concurrently added or removed without the existing bit sync blocks there is today in queueimpl

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java:
```diff
@@ -2409,14 +2410,10 @@ private void deliver() {
 
       long timeout = System.currentTimeMillis() + DELIVERY_TIMEOUT;
-
+      consumers.reset();
```
if any exception would be thrown before it is ok that `reset` won't be called? 
If not, better to wrap the whole logic with ìtry..finally` and `reset`

michaelandrepearce replied:
Its fine. We actually only want to reset on a succesful handled nomatch or expired. Or before starting iterating

franz1981 commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java:
```diff
@@ -2497,45 +2484,36 @@ private void deliver() {
+                  removeMessageReference(holder, ref);
+                  handledconsumer = consumer;
                   handled++;
+                  consumers.reset();
```
if `removeMessageReference` would throw any exeption is fine to have `consumers.reset` not called?
if not, uses `try...finally`

michaelandrepearce replied:
Yes its fine. Would have no negative effect, and actually have same behaviour as old.

franz1981 commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/PriorityCollection.java:
```diff
@@ -0,0 +1,319 @@
+   }
+
+   @Override
+   public synchronized boolean add(T t) {
```
Looking to the usage on `QueueImpl` it seems that is being always called into a `synchronized (this) {` block ie it doesn't need to be `synchronized` and same same for `remove`. 
If we allow `snapshotting` from a different thread is fine, but I would document it instead.
I know that's a less safe approach than putting `sychronized` all around, but I do believe that we would have to do the same on  `LinkedListImpl` for this same reason if we need to `protect` correct usage at any cost.

michaelandrepearce replied:
This needs to be concurrent safe on modifications

franz1981 commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/SessionCreateConsumerMessage.java:
```diff
@@ -113,6 +133,7 @@ public int hashCode() {
       result = prime * result + (browseOnly ? 1231 : 1237);
       result = prime * result + ((filterString == null) ? 0 : filterString.hashCode());
       result = prime * result + (int) (id ^ (id >>> 32));
+      result = prime * result + priority;
```
I would uses `Integer::hashCode(priority)` that would `agitate` the value

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpReceiverPriorityTest.java:
```diff
@@ -0,0 +1,144 @@
+
+
+      for (int i = 0; i < 5; i++) {
+         AmqpMessage message1 = receiver1.receiveNoWait();
```
I believe the test clients receiveNoWait only polls its local queue, so this might now likely sporadically fail due to racing the deliveries. I was only suggesting recieveNoWait as potential initial verification within the loop for non-receiving consumers, to be followed up or substituted by a more stringent final check.

michaelandrepearce commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/SessionCreateConsumerMessage.java:
```diff
@@ -84,6 +94,10 @@ public void setFilterString(SimpleString filterString) {
       this.filterString = filterString;
    }
 
+   public void setPriority(byte priority) {
```
Keeping inline with the rest of the fields and general approach. Id rather keep to status quo

michaelandrepearce commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/PriorityCollection.java:
```diff
@@ -0,0 +1,319 @@
+      PriorityHolder<T>[] current = getArray();
+      int size = 0;
+      for (PriorityHolder<T> priorityHolder : current) {
+         size += priorityHolder.getValues().size();
```
If we have a case of the number of active consumers on a single queue is integer max value we will have many areas to address tbh! And that is some super extreme case.

We can reject on adding a new consumer if adding the new consumer would breach int max value, if we are really worried

michaelandrepearce replied:
If we have a case of the number of active consumers on a single queue is integer max value we will have many areas to address tbh! And that is some super extreme case.

We can reject on adding a new consumer if adding the new consumer would breach int max value, if we are really worried

michaelandrepearce commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/PriorityCollection.java:
```diff
@@ -0,0 +1,319 @@
+      int high = len - 1;
+
+      while (low <= high) {
+         int mid = (low + high) >>> 1;
```
This is just coding pref for this stuff im happier with shift operations. Its how alot of the jvm classes do this.

franz1981 replied:
I know but the reason is because some optimisations were not present in old JDK version and the classes aren't changed...

michaelandrepearce replied:
still just a coding preference, i see you even do bitwise alot when can be just written in standard math ;) ..... 

anyhow have changed to avoid disagreement.

michaelandrepearce commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/MultiResettableIterator.java:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @param <T> type of the class of the iterator.
+ */
+public class MultiResettableIterator<T> extends MultiIterator<T> implements ResettableIterator<T> {
```
You could but this then makes using MutiIterator ugly as hell with little gain

michaelandrepearce commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/MultiIterator.java:
```diff
@@ -0,0 +1,78 @@
+ *
+ * @param <T> type of the class of the iterator.
+ */
+public class MultiIterator<T> implements Iterator<T> {
```
Yeah i had a version like that but it makes using simple MultiIterator ugly to use. Uneededly. And to extend if needed only needs a cast. In case of resetable.

michaelandrepearce replied:
OK, so to do this and keep the cleaner user interface what i can do is make an abstract base which has most of this logic in, and have the more ugly generics. And then make MultiIterator extend that which then keeps its cleaner generics, and like wise MultiResettableIterator. Ill update PR to see if this is good compromise, or if you'd prefer me to revert and simply keep the cast i had originally.

michaelandrepearce commented:
@gemmellr ok will reduce back down the extra tests i added. To those you suggest. 

Also re activemq5. Fair enough, ill reduce / alter this then.

michaelandrepearce commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/MultiResettableIterator.java:
```diff
@@ -0,0 +1,44 @@
+   protected void moveTo(int index) {
+      super.moveTo(index);
+      if (index > -1) {
+         ((ResettableIterator<T>) get(index)).reset();
```
It could but changing MultiIterators generics then makes it ugly for any use of it. Id rather ensure usage of the classes is kept clean over their internals. And a single cast isnt a biggie imo for the benefit of a cleaner usage by the end using code that uses it.

michaelandrepearce commented on artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/MultiIterator.java:
```diff
@@ -0,0 +1,78 @@
+public class MultiIterator<T> implements Iterator<T> {
+
+   private final Iterator<T>[] iterators;
+   int index = -1;
```
Agreed

michaelandrepearce commented:
@franz1981 good point around adding further tests to the extracted out bits. Agree it will make everything more robust

michaelandrepearce commented:
@gemmellr pushed changes based on latest comments, thanks.

michaelandrepearce commented:
@franz1981 pushed following changes based on your comments that agree with, for others i have left comment for us to discuss. 

Changes:
1) Ensured total size of the priority collection can never exceed Integer.MAX_VALUE, by ensuring this on add, thus the edge case you were worried about of the calcSize being greater than int cannot occur. Note if someone has that many consumers, we probably want to have some discussions with them as they would be some power user!!! ;)

2)  Avoid double volatile read of changedIterator in reset method.

3) Removed need for a cast in MultiResettableIterator

gemmellr commented:
@ Thanks. I'd still rather some of the tests weren't burning 250ms various times waiting to test something that can be verified 'immediately', but its a lot better.

I'm still not sure if you've seen this comment (github is hiding it as 'resolved') so I'll copy it here:

https://github.com/apache/activemq-artemis/pull/2490/files#r246398611
> Does this comment cover to the bit around the safety of always sending the new additional data even to old servers? I can't tell if its covered.
>
> I think it should at the very least be commented what/when the encoding+decoding handling behaviour changed so folks can understand the implications later without heading to find past commits.


which was related to part of an earlier comment https://github.com/apache/activemq-artemis/pull/2490/files#r245955337), mostly:
> Also, is the reverse case safe, does an older server failing to read the additional value (seemingly always sent now) have potential to lead to any issues on older servers, i.e how might the buffer continue to be used later if at all? Should the client omit the value for older servers? (Or does the presumed version change prevent the new client working with the old server anyway? I don't know how that stuff is handled, just commenting from reading the diff here).

michaelandrepearce commented:
@gemmellr so this is the same pattern used quite a bit, e.g. if you look at in master: ClusterTopologyChangeMessage_V2 it was used when adding backupGroupName, like wise you will see similar in other XXXMessage.

By the extra bits only being at the end, older servers simply read the old fields out it just doesnt read the extra byte. And having the readableBytes check means if the reverse where an older client wont have provided the extra bytes, then this is where we default it.

There is also a bunch on compatibility tests that @clebertsuconic made, that if there was a compatibility break it would be picked up.

gemmellr commented:
@michaelandrepearce I'd have made the same comment for those changes if reviewing them, as I would probably have to go digging through commits now if I wanted to understand the evolution there for subsequent changes, when at least a simple comment might have done the trick. Giving up on this point though, enough time wasted looking to make things clearer for later readers to avoid people needing to waste time :)

Its more the other side of the comment that primarily concerns me, that the new client always sends the extra detail to old servers despite knowing they cant use it. The fact the old server then doesnt read out the extra info in the buffer is what I find most concerning, as the data is going to be there in the buffer, and so how that buffer is later used by the old server becomes important following the changes even though it doesnt support the feature. The basic compatibility tests dont really ease that concern. It just seems simpler and safer to ensure ongoing compatibility isnt affected by not sending old servers new data we know they cant use, but could perhaps trip over in unexpected ways.

Looking at the existing code now rather than just the diff here, it seems the very first thing it does after decoding is use the current buffer index as a 'size', which then might not be accurate any more depending on how its being used or how it interacts with other areas later. I haven't traced further usage of the buffer or that size, but I personally wouldn't be comfortable with the new data being sent to old servers based on that alone. Who knows what things other areas and even older different versions of the code might do in the face of changing the buffer content.

michaelandrepearce commented:
@gemmellr as i stated, this approach has been used many times over, historically and is obviosuly proven, this is not something i invented, im just following the same working pattern.

gemmellr commented:
Sometimes tests don't pick up issues that may exist and we later find issues prompting more specific tests. You essentially don't need to rely on the compatibility test in cases like this though if you simply don't change the behaviour. Given the old servers are known not to support the feature, changing behaviour against them when not needed them doesn't seem beneficial at all, but could be detrimental. 

If an apparent issue on the first line executed after failing to read new-but-unusable-data sent to an old servers isn't enough reason for pause here, I'm don't know what would be.

michaelandrepearce commented:
@gemmellr a created a new SessionCreateConsumerMessage_V2 so hopefully this addresses this and added a version check on channel.

also addressed merge conflicts after recent merges to master from other prs.

gemmellr commented:
I'd have been content with an if and comment in the original class, but happy with the class if that is convention.

michaelandrepearce commented:
Oh, ok ill try do that.

michaelandrepearce commented:
Ok, I put back to original class, and added an if check to it, with comments.

Squashed everything also.

michaelandrepearce commented:
@franz1981 i know i still owe you those tests...

michaelandrepearce commented:
@franz1981 added the tests i owed you. Sorry took some time, had the weekend off.

michaelandrepearce commented:
@franz1981 good to go?

clebertsuconic commented:
Great work here!


I'm running testsuite just to make sure everything is covered.

clebertsuconic commented:
I tried pulling your branch, but the conflict that I thought I fixed caused a compilation error.


do you mind rebasing it please?

michaelandrepearce commented:
rebased

clebertsuconic commented:
one test failure at least: JMSMessageGroupsTest


The testsuite didn't complete.. but that one is failing for sure.

clebertsuconic commented:
I will spend some time understanding the failure and propose changes. I will communicate through here.. and if you find anything please let me know here as well.

michaelandrepearce commented:
Ill have a look

michaelandrepearce commented:
@clebertsuconic ok so i know whats going on, another feature i have implemented for 2.7.0 reset of message groups, the test i had created is a little over strict and tested exact current behaviour, rather than what is just needed, that when resetting a message group, the selection on the next consumer for that group would have been the next consumer that joined. Where as actually when you reset the group in reality you just open the oppurtunity for another consumer to take the group.

michaelandrepearce commented:
Sorted locally just extracting repeatble iterator logic as franz is keen on this

michaelandrepearce commented:
@clebertsuconic sorted,

just added ability in consumers to repeat (new repeatable iterator), and then if group, we tell it to repeat, that provides the original behavior, and test is passing locally.

clebertsuconic commented:
Current list of failures:


Test Name | Duration | Age
-- | -- | --
org.apache.activemq.artemis.tests.integration.client.AutogroupIdTest.testGroupIdAutomaticallySetMultipleProducers | 37 ms | 1
org.apache.activemq.artemis.tests.integration.client.MessageGroupingTest.testMultipleGroupingXACommit | 47 ms | 1
org.apache.activemq.artemis.tests.integration.client.MessageGroupingTest.testLoadBalanceGroups | 21 ms | 1
org.apache.activemq.artemis.tests.integration.client.MessageGroupingTest.testMultipleGroupingTXCommit | 0.25 sec | 1
org.apache.activemq.artemis.tests.integration.client.MessageGroupingTest.testMultipleGroupingConsumeHalf | 0.79 sec | 1
org.apache.activemq.artemis.tests.integration.client.MessageGroupingTest.testMultipleGroupingXARollback | 55 ms | 1
org.apache.activemq.artemis.tests.integration.client.MessageGroupingTest.testMultipleGrouping | 14 ms | 1
org.apache.activemq.artemis.tests.integration.client.MessageGroupingTest.testMultipleGroupingTXRollback | 0.28 sec | 1
org.apache.activemq.artemis.tests.integration.cluster.failover.ReplicatedMultipleServerFailoverExtraBackupsTest.testStartLiveFirst | 1 min 8 sec | 1
org.apache.activemq.artemis.tests.integration.jms.client.GroupingTest.testManyGroups | 6 sec | 1
org.apache.activemq.artemis.tests.integration.jms.client.GroupingTest.testManyGroups | 6 sec | 1
org.apache.activemq.artemis.tests.integration.jmx.JmxConnectionTest.testJmxConnection | 1.1 sec | 1
org.apache.activemq.artemis.tests.integration.journal.AIOJournalImplTest.testDoubleDelete | 53 ms | 1
org.apache.activemq.artemis.tests.integration.journal.AIOUnbuferedJournalImplTest.testDoubleDelete | 39 ms | 1
org.apache.activemq.artemis.tests.integration.journal.MappedJournalImplTest.testDoubleDelete | 39 ms | 1
org.apache.activemq.artemis.tests.integration.journal.MappedUnbuferedJournalImplTest.testDoubleDelete | 43 ms | 1
org.apache.activemq.artemis.tests.integration.journal.NIOJournalImplTest.testDoubleDelete | 36 ms | 1
org.apache.activemq.artemis.tests.integration.journal.NIONoBufferJournalImplTest.testDoubleDelete | 18 ms | 1
org.apache.activemq.artemis.tests.integration.mqtt.MqttAcknowledgementTest.testAcknowledgementQOS1 | 5.6 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTFQQNTest.testSendAndReceiveMQTTSpecial1 | 11 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTOpenwireTest.testWildcards | 3.2 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTSecurityCRLTest.crlNotRevokedTest | 5.2 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAndReceiveAtLeastOnce | 5.3 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testManagementQueueMessagesAreAckd | 5.2 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testMQTTRetainQoS | 7.3 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSubscribeMultipleTopics | 5 min 0 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAndReceiveExactlyOnceWithInterceptors | 6 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAtLeastOnceReceiveAtMostOnce | 5.4 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testNoMessageReceivedAfterUnsubscribeMQTT | 5.4 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testMQTTPathPatterns | 5.3 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testUniqueMessageIds | 5.3 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAndReceiveRetainedMessages | 5.4 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testRetainedMessage | 5.9 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testResendMessageId | 5.3 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testClientConnectionFailure | 11 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAtMostOnceReceiveExactlyOnce | 5.5 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testAnycastAddressWorksWithMQTT | 3.5 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testAnycastPrefixWorksWithMQTT | 2.4 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testReceiveMessageSentWhileOffline | 5.4 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testAmbiguousRoutingWithMQTT | 3.2 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testLinkRouteAmqpReceiveMQTT | 5.6 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testPacketIdGeneratorNonCleanSession | 2.1 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testWillMessageIsRetained | 11 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAndReceiveAtMostOnce | 5.7 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAtLeastOnceReceiveExactlyOnce | 7.9 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testDuplicateSubscriptions | 7.6 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendJMSReceiveMQTT | 11 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testCleanSessionForSubscriptions | 11 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testRetainedMessagesAreCorrectlyFormedAfterRestart | 7.7 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testUnsubscribeMQTT | 23 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAndReceiveLargeMessages | 6.4 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testClientConnectionFailureSendsWillMessage | 15 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MQTTTest.testSendAndReceiveMQTT | 12 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MqttClusterRemoteSubscribeTest.unsubscribeRemoteQueue | 10 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MqttClusterRemoteSubscribeTest.unsubscribeExistingQueue | 10 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MqttClusterRemoteSubscribeTest.unsubscribeRemoteQueueMultipleSubscriptions | 8.7 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MqttClusterRemoteSubscribeTest.unsubscribeRemoteQueueWildCard | 10 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MqttClusterWildcardTest.loadBalanceRequests | 36 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.MqttClusterWildcardTest.wildcardsWithBroker1Disconnected | 18 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.PahoMQTTTest.testLotsOfClients | 39 sec | 1
org.apache.activemq.artemis.tests.integration.mqtt.imported.PahoMQTTTest.testSendAndReceiveMQTT | 1 min 2 sec | 1
org.apache.activemq.artemis.tests.integration.openwire.cluster.TemporaryQueueClusterTest.testClusteredQueue | 7.7 sec | 1
org.apache.activemq.artemis.tests.integration.openwire.cluster.TemporaryQueueClusterTest.testTemporaryQueue | 3.9 sec | 1
org.apache.activemq.artemis.tests.integration.paging.GlobalPagingTest.testDeletePhysicalPages[storeType=FILE, mapped=true] | 2.9 sec | 1
org.apache.activemq.artemis.tests.integration.paging.GlobalPagingTest.testReceiveImmediate[storeType=FILE, mapped=true] | 2.1 sec | 1
org.apache.activemq.artemis.tests.integration.paging.PageCountSyncOnNonTXTest.testSendNoTx | 57 sec | 1
org.apache.activemq.artemis.tests.integration.paging.PagingTest.testPurge[storeType=FILE, mapped=false] | 34 sec | 1
org.apache.activemq.artemis.tests.integration.plugin.MqttPluginTest.testSendAndReceiveMQTT | 15 sec | 1
org.apache.activemq.artemis.tests.integration.server.ExpiryRunnerTest.testExpireConsumeHalf | 2.5 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompConnectionCleanupTest.sendSTOMPReceiveMQTT[ws+v10.stomp] | 11 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompConnectionCleanupTest.sendSTOMPReceiveMQTT[tcp+v10.stomp] | 11 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompConnectionCleanupTest.testMultiProtocolConsumers[tcp+v10.stomp] | 3.6 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompTest.sendSTOMPReceiveMQTT[ws+v10.stomp] | 10 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompTest.sendSTOMPReceiveMQTT[tcp+v10.stomp] | 10 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompWebSocketMaxFrameTest.testStompSendReceiveWithMaxFramePayloadLength[ws+v10.stomp] | 14 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompWebSocketMaxFrameTest.testStompSendReceiveWithMaxFramePayloadLength[ws+v11.stomp] | 13 sec | 1
org.apache.activemq.artemis.tests.integration.stomp.StompWebSocketMaxFrameTest.testStompSendReceiveWithMaxFramePayloadLength[ws+v12.stomp] | 15 sec | 1

michaelandrepearce commented:
@clebertsuconic can you recheck this against master, i know things like testDoubleDelete are failing currently on master without this PR.

michaelandrepearce commented:
@clebertsuconic could you look at this https://github.com/apache/activemq-artemis/commit/b3f0a87e0db4832585a0460633889e13ef4bcbf8#r31982567, all builds are failing since that was merged to master.....

michaelandrepearce commented:
@clebertsuconic fully rebased (including over the FQQN for producers stuff)

michaelandrepearce commented:
@clebertsuconic so as noted, the message groups ones is sorted, repeat needs to occur only if groupconsumer exist. 

It does seem MQTTTest and StompWebSocketMaxFrameTest is directly broken due to this, but i cant think to why, any ideas welcome, ill continue to look into

michaelandrepearce commented:
@clebertsuconic ive found the pesky thing, in ServerSessionImpl where i had added a new parameter to the createConsumer method, the old method signature where we pass though but default the priority in the new, i had missed forwarding on the credits value, thus the issue. 

Typical one liner causing lots of headache ;)

michaelandrepearce commented:
@clebertsuconic  hows this looking now, we good to merge???

clebertsuconic commented:
Let me run one final round of testsuite.  If all good will merge it tomorrow.

clebertsuconic commented:
It’s important to keep the testsuite clean.  As we should release soon.

clebertsuconic commented:
@michaelandrepearce I'm pretty sure I will be able to merge this tomorrow. Testsuite had some unrelated failures. I just want to double check.. if there's any issues I'm sure i will be able to figure out.


will let you know tomorrow. I got the results pretty late after I finished. leave it with me for tomorrow first thing please?

clebertsuconic commented:
I added one change on top of yours during merge... to remove unused variables.. 


merged!


you are the man!

michaelandrepearce commented:
@clebertsuconic some of those variables are used but accesed by atomic updaters. Have commented on your commit
